elastic/kibana
# [E&C][ES Query] adds runtime mappings and fields support to the ES Query ruletype

## Summary

This PR adds Runtime Fields support to the ES Query Rule Type when using the DSL Query mode.

By adding support for `runtime_mappings` and `fields` parameters we enable customers to:

1. enrich their `hits` with custom field formatting
2. enrich their `hits` with custom runtime fields
3. address a common pain point where data isn't always ingested in an ideal manner given the output necessary for their preferred actions

Docs preview [here](https://kibana_138427.docs-preview.app.elstc.co/guide/en/kibana/master/rule-type-es-query.html).

### How to validate

You can now define rules with runtime fields, suc has:
```json
{
  "runtime_mappings": {
    "day_of_week": {
      "type": "keyword",
      "script": {
        "source": "emit(doc['@timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
      }
    },
    "kebabAsSnake": {
      "type": "keyword",
      "script": {
        "source": "emit(doc['kibana.alert.rule.rule_type_id'].value.replace(\"-\",\"_\"));"
      }
    }
  },
  "fields": ["kebabAsSnake", "day_of_week"],
  "query":{
    "match_all" : {}
  }
}
```

and then use these in actions like so:
```
{{#context.hits}}
timestamp: {{_source.@timestamp}}
dow: {{fields.day_of_week}}
kebabAsSnake: {{fields.kebabAsSnake}}
{{/context.hits}}
```

Validating this requires configuring a variety of runtime field types and making sure these are passed into the `hits` context variable correctly.
I'm not sure I'm familiar enough with runtime fields to cover all formats, so feel free to push the limits of what you know is possible - I might have missed something, but in practice these are passed "as-is" from the ES result to the action, so I'd be surprised if anything doesn't work as expected here.

# Release Note

The ES Query Rule Type now supports Runtime Mappings and the Fields parameters when using an Elasticsearch DSL query.
Any rule created prior to 8.5 is migrated to strip out any legacy usage of these parameters (which would have been ignored prior to 8.5.0) to ensure they do not impact those rules' behaviour after upgrade.

### Checklist

Delete any items that are not applicable to this PR.


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
